torchdet3d/dataloaders/objectron_main.py

```diff
@@ -0,0 +1,156 @@
+            draw_kp(transformed_image.numpy(), transformed_keypoints.numpy(), 'image_after_pipeline.jpg')
+
+        if self.mode == 'test':
+            return (image,
+                    torch.from_numpy(transformed_image),
+                    transformed_keypoints,
+                    category)
+
+        return transformed_image, transformed_keypoints, category
+
+    def crop(self, image, keypoints):
+        ''' fetch 2D bounding boxes from keypoints and crop the image '''
+        real_h, real_w, _ = image.shape
+
+        # clamp bbox coordinates according to image shape
+        clipped_bb = self.clip_bb(keypoints, real_w, real_h)
+        # crop 2D bounding box from image by given 3D keypoints
+        # (min x, miny) - left lower corner;
+        # (max x, max y) - upper right corner
+        x0 = self.clamp(min(clipped_bb[:,0]) - 10, 0, real_w)
+        y0 = self.clamp(min(clipped_bb[:,1]) - 10, 0, real_h)
+        x1 = self.clamp(max(clipped_bb[:,0]) + 10, 0, real_w)
+        y1 = self.clamp(max(clipped_bb[:,1]) + 10, 0, real_h)
+
+        # prepare transformation for image cropping and kp shifting
+        transform_crop = A.Compose([
+                        A.Crop(x0,y0,x1,y1),
+                        ], keypoint_params=A.KeypointParams(format='xy'))
+        # do actual crop and kp shift
+        transformed = transform_crop(
+                                image=image,
+                                keypoints=clipped_bb
+                                )
+
+        crop_img = transformed['image']
+        bb = transformed['keypoints']
+        assert len(bb) == 9
+
+        return bb, crop_img
+
+    def clip_bb(self, bbox, w, h):
+        ''' clip offset bbox coordinates
+        bbox: np.array, shape: [9,2], repr: [x,y]'''
+        clipped_bbox = np.empty_like(bbox)
+        clamped_x = list(map(lambda x: self.clamp(x, 3, w-3), bbox[:,0]))
+        clamped_y = list(map(lambda y: self.clamp(y, 3, h-3), bbox[:,1]))
+        clipped_bbox[:,0] = clamped_x;  clipped_bbox[:,1] = clamped_y
+        return clipped_bbox
+
+    @staticmethod
+    def clamp(x, min_x, max_x):
+        return min(max(x, min_x), max_x)
+
+
+def test():
+    "Perform dataloader test"
+    def super_vision_test(root, mode='val', transform=None, index=7):
+        ds = Objectron(root, mode=mode, transform=transform, debug_mode=True)
+        _, bbox, _ = ds[index]
+        assert bbox.shape == (9,2)
+
+    def dataset_test(root, mode='val', transform=None, batch_size=5):
+        ds = Objectron(root, mode=mode, transform=transform)
+        dataloader = DataLoader(ds, batch_size=batch_size, shuffle=True)
+        iter_dt = iter(dataloader)
+        img_tensor, bbox, cat = next(iter_dt)
+        ic(mode)
+        ic(cat)
+        ic(img_tensor.shape)
+        ic(bbox.shape)
+        assert img_tensor.shape == (batch_size, 3, 290, 290)
+        assert bbox.shape == (batch_size, 9, 2)
+
+    root = 'data'
+    transform = A.Compose([ ConvertColor(),
```

Comment:
ConvertColor should be right before ToTensor. Early channel swapping cause misinterpreting colors by the other opencv functions you may apply after as augmentations.

Reply:
What about color transformation? Wouldn't it distort color spaces and do expected augmentation in the wrong way?

Reply:
This is exactly that happening now: all the opencv's color transforms assume BGR layout whereas you convert the image to RGB from the beginning.

Reply:
according to the Albumentation documentation and their examples, an augmentation pipeline should receive RGB format.
Please, refer to this [paragraph](https://albumentations.ai/docs/getting_started/image_augmentation/#step-3-read-images-from-the-disk)
So, I suggest keeping ConverColor first